contracts/BucketAuction.sol

```diff
@@ -0,0 +1,282 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.4;
+
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+import "@openzeppelin/contracts/access/AccessControl.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/token/common/ERC2981.sol";
+import "erc721a/contracts/ERC721A.sol";
+
+import "./IBucketAuction.sol";
+import "./ERC721M.sol";
+
+contract BucketAuction is IBucketAuction, ERC721M, ReentrancyGuard {
+    uint256 private _minimumContributionInWei;
+    uint256 private _price;
+    bool private _auctionActive;
+    mapping(address => User) private _userData;
+
+    constructor(
+        string memory collectionName,
+        string memory collectionSymbol,
+        string memory tokenURISuffix,
+        uint256 maxMintableSupply,
+        uint256 globalWalletLimit,
+        address cosigner,
+        uint256 minimumContributionInWei
+    )
+        ERC721M(
+            collectionName,
+            collectionSymbol,
+            tokenURISuffix,
+            maxMintableSupply,
+            globalWalletLimit,
+            cosigner
+        )
+    {
+        _minimumContributionInWei = minimumContributionInWei;
+    }
+
+    modifier isAuctionActive() {
+        if (!_auctionActive) revert BucketAuctionNotActive();
+        _;
+    }
+
+    modifier isAuctionInactive() {
+        if (_auctionActive) revert BucketAuctionActive();
+        _;
+    }
+
+    function getMinimumContributionInWei() external view returns (uint256) {
+        return _minimumContributionInWei;
+    }
+
+    function getPrice() external view returns (uint256) {
+        return _price;
+    }
+
+    function getAuctionActive() external view returns (bool) {
+        return _auctionActive;
+    }
+
+    function getUserData(address user) external view returns (User memory) {
+        return _userData[user];
+    }
+
+    /**
+     * @notice begin the auction.
+     * @dev cannot be reactivated after price has been set.
+     * @param b set 'true' to start auction, set 'false' to stop auction.
+     */
+    function setAuctionActive(bool b) external onlyOwner cannotMint {
+        if (_price != 0) revert PriceHasBeenSet();
+        _auctionActive = b;
+    }
+
+    /**
+     * @notice place a bid in ETH or add to your existing bid. Calling this
+     *   multiple times will increase your bid amount. All bids placed are final
+     *   and cannot be reversed.
+     */
+    function bid() external payable isAuctionActive nonReentrant cannotMint {
+        User storage bidder = _userData[msg.sender]; // get user's current bid total
+        uint256 contribution_ = bidder.contribution; // bidder.contribution is uint216
+        unchecked {
+            // does not overflow
```

Comment:
Why do we use uint216 for contribution and why it doesn't overflow?

Reply:
compact the User struct storage

Reply:
TODO: let's test if using 216 helps